File: library/src/test/java/io/fotoapparat/log/FileLoggerTest.java
package io.fotoapparat.log;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.io.File;
import java.nio.charset.Charset;

import static junit.framework.Assert.assertEquals;
import static org.apache.commons.io.FileUtils.readFileToString;

public class FileLoggerTest {

    static final File FILE = new File("test");

    FileLogger testee;

    @Before
    public void setUp() throws Exception {
        ensureFileDeleted();

        testee = new FileLogger(FILE);
    }

    @Test
    public void writeMessage() throws Exception {
        // When
        testee.log("message");
        testee.log("message");

        // Then
        assertEquals(
                "message\nmessage\n",
                readFileToString(FILE, Charset.defaultCharset())
        );
    }

    @Test
    public void ignoreExceptions() throws Exception {
        // Given
        FileLogger logger = new FileLogger(new File("/"));

        // When
        logger.log("message");

        // Then
        // No exception is thrown
    }

    @After
    public void tearDown() throws Exception {
        ensureFileDeleted();
    }

    private void ensureFileDeleted() {
        if (FILE.exists() && !FILE.delete()) {
            throw new IllegalStateException("Test file still exists: " + FILE);
        }
    }

}

File: fotoapparat/src/test/java/io/fotoapparat/log/FileLoggerTest.kt
package io.fotoapparat.log

import org.junit.After
import org.junit.Before
import org.junit.Test
import java.io.File
import kotlin.test.assertEquals

internal class FileLoggerTest {

    internal val file = File("test")
    lateinit var testee: FileLogger

    @Before
    fun setUp() {
        file.ensureFileDeleted()

        testee = FileLogger(file)
    }

    @Test
    fun `Write message`() {
        // When
        testee.log("message")
        testee.log("message")

        // Then
        assertEquals(
                "message\nmessage\n",
                file.readText()
        )
    }

    @Test
    fun `Ignore exceptions`() {
        // Given
        val logger = FileLogger(File("/"))

        // When
        logger.log("message")

        // Then
        // No exception is thrown
    }

    @After
    fun tearDown() {
        file.ensureFileDeleted()
    }

}

private fun File.ensureFileDeleted() {
    if (exists() && !delete()) {
        throw IllegalStateException("Test file still exists: " + this)
    }
}
